### Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.

In [15]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor

In [16]:
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

X_train = pd.read_pickle('X_train.pkl')
X_test = pd.read_pickle('X_test.pkl')
y_train = pd.read_pickle('y_train.pkl')
y_test = pd.read_pickle('y_test.pkl')

In [17]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

In [18]:
y_pred = model.predict(X_test)

In [19]:
check_test = pd.DataFrame({
    "y_test": y_test['price'],
    "y_pred": y_pred.flatten()
})
check_test.head()

y_test     y_pred
173    23.6  22.806412
274    32.4  31.131464
491    13.6  16.339125
72     22.8  23.810726
452    16.1  17.139521

In [20]:
from sklearn.metrics import r2_score

In [21]:
r2 = r2_score(check_test['y_test'], check_test['y_pred'])
r2

0.87472606157312

Напишите в комментариях к коду, какая модель в данном случае работает лучше.

**В данном случае, модель линейной регрессии работает хуже (r2 = 0.7112260057484908), чем модель случайного леса (r2 = 0.87472606157312)**

### Задание 3

Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.

In [22]:
from sklearn.ensemble import RandomForestRegressor

С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [30]:
model.feature_importances_

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [31]:
fis = pd.DataFrame({
    'name': X_train.columns,
    'feature_importance': model.feature_importances_
})
fis

name  feature_importance
0      CRIM            0.031676
1        ZN            0.001543
2     INDUS            0.007138
3      CHAS            0.001236
4       NOX            0.014269
5        RM            0.402682
6       AGE            0.014299
7       DIS            0.063973
8       RAD            0.005281
9       TAX            0.011525
10  PTRATIO            0.018081
11        B            0.012451
12    LSTAT            0.415847

In [32]:
fis.nlargest(2, 'feature_importance')

name  feature_importance
12  LSTAT            0.415847
5      RM            0.402682

**Как видно выше, самые значимые признаки: LSTAT, RM**